### PR TITLE
RP-183: Add XML helper methods to script framework

### PR DESCRIPTION
**Pull request by gregcharles**
This is the first phase toward creating a basic DSL for the script pipeline, and I've included a README.md to demonstrate some samples of how the scripts for XML transformation could be written. I'd appreciate any feedback on the approach and general and more specific design elements. E.g., does it make sense to dynamically add XML methods to the script class or would it be better to have separate base classes based on the type of  input being processed? Also, are there other helper methods that we could add?

### CODE REVIEW

File: script-common/src/test/resources/xsl/exam.xsl [hunk 1]
```diff
@@ -0,0 +1,174 @@
+<!DOCTYPE xsl:stylesheet >
+<xsl:stylesheet
```
Comment (esotrope) [inline script-common/src/test/resources/xsl/exam.xsl]: are the xsl used to test feature parity? or will it be an option to use them in the pipeline?

File: script-common/README.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
+## RDW Ingest - Script Pipeline
+
+### Helper functions for XML
+XML support can be added to a Script class by calling:
+ ```script.addXmlExtensions()```. We assume that an XML document
+ will be converted into another XML document, and all elements, including
+ commentsnot, not specifically altered, should appear in the converted
```
Comment (malaffoon) [inline script-common/README.md]: nit: typo

File: script-common/README.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
+## RDW Ingest - Script Pipeline
+
+### Helper functions for XML
+XML support can be added to a Script class by calling:
+ ```script.addXmlExtensions()```. We assume that an XML document
```
Comment (malaffoon) [inline script-common/README.md]: nit: when inlining mono formatting you can surround with single back-ticks. I'd give an example, but this comment is also markdown and i'm too lazy to figure out how to suppress the formatting in the example.

File: script-common/README.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ ```groovy
+transform '//Movie' by {  
+    if (it.title == 'Jaws') {
+        it.title = it.title.toUpperCase()
+    }
+}
+```
+
+The syntax ```movie.title``` (or ```it.title```) refers to the title
+attribute of the movie element. It is shorthand for ```movie.getAttributeValue()```,
```
Comment (malaffoon) [inline script-common/README.md]: The `movie.getAttributeValue()` example is confusing. Should it be `movie.getAttributeValue('title')`?

File: script-common/README.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+```
+
+It is also possible to reverse this logic to delete all elements
+(matching the XPath expression) that don't match the rule:
+ 
+```groovy
+delete '//Movie' unless { movie -> 
+    movie.actor == 'Richard Basehart'
+}
+```
```
Comment (malaffoon) [inline script-common/README.md]: Is there nice syntax for applying a transform while using when/unless?

File: script-common/src/main/groovy/org/opentestsystem/rdw/ingest/script/DSLScriptBase.groovy [hunk 6]
```diff
@@ -19,7 +34,116 @@ abstract class DSLScriptBase extends Script {
...
+     * @return reference to the script class (for chained calls).
+     */
+    DSLScriptBase enableXmlExtensions() {
+        Document document = new SAXBuilder().build(getProperty("input"))
+
+        this.setProperty("document", document)
+
+        // Unescapes up embedded XML in TRT values so it can treated as regular XML.
+        String.metaClass.unescapeXml {
+            delegate.replaceAll('&lt;','<').replaceAll('&gt;','>')
```
Comment (malaffoon) [inline script-common/src/main/groovy/org/opentestsystem/rdw/ingest/script/DSLScriptBase.groovy]: Is there not a helper in the XML libraries for this? I ask because there are additional sequences that need to be unescaped, aren't there?

File: script-common/src/test/java/org/opentestsystem/rdw/ingest/script/ScriptXmlTest.java [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+    public void itShouldFilterOutElements() {
+        // Test ability to filter out elements with a boolean rule. All elements matching the XPath expression
+        // and also matching this rule will be removed.
+        String scriptDefinition =
+                "delete '//Item' when { item ->" + LF +
+                "    item.key == '37115' || item.key == '92135'" + LF +
+                        "}" + LF +
+                "outputXml";
+
+        Assert.assertThat(originalXml, hasXPath("//Item[@key='37115']", not(isEmptyOrNullString())));
```
Comment (malaffoon) [inline script-common/src/test/java/org/opentestsystem/rdw/ingest/script/ScriptXmlTest.java]: strongly recommend statically importing `assertThat`

File: script-common/build.gradle [hunk 8]
```diff
@@ -15,5 +15,10 @@ dependencies {
     compile 'org.codehaus.groovy:groovy-all'
     compile 'org.jdom:jdom2:2.0.5'
     compile 'jaxen:jaxen:1.1.4'
+    compile 'net.sf.saxon:Saxon-HE'
+
+    testCompile 'org.xmlunit:xmlunit-core'
+    testCompile 'org.xmlunit:xmlunit-matchers'
+    testCompile 'org.springframework.boot:spring-boot-starter-test'
```
Comment (malaffoon) [inline script-common/build.gradle]: curious why you'd need spring-boot for all this

File: script-common/src/test/resources/xsl/exam.xsl (hunk 1 above)
Comment (gregcharles) [reply script-common/src/test/resources/xsl/exam.xsl]: The xsl can be called from the script in place of the coded rules, or to complement them. I do have one test case that performs the complete transformation by applying this xsl file via a Groovy script, and another one that uses a reduced set of xsl transformations and makes up the rest with script commands.

File: script-common/README.md (hunk 2 above)
Comment (gregcharles) [reply script-common/README.md]: Oh, right. "comment snot" should have a space in it. I'll fix that!

File: script-common/README.md (hunk 3 above)
Comment (gregcharles) [reply script-common/README.md]: OK, good to know! I'm changing these.

File: script-common/README.md (hunk 4 above)
Comment (gregcharles) [reply script-common/README.md]: It's just wrong. I'll fix that. 

You're right that Groovy usually maps movie.title to movie.getTitle(). However, I have set it up to override that behavior, and instead call getAttributeValue('propertyname'). I sort of wonder if I shouldn't also get sub-elements with the same syntax. If an attribute name collided with a child element name though, it would get messy.

File: script-common/README.md (hunk 5 above)
Comment (gregcharles) [reply script-common/README.md]: You'd want to transform something inside a delete command? Do you mean to transform the ones you don't delete?  You can actually do any of the transformations here that you can do in a transform command. The key is that you have to return true or false, which determines whether the element will be kept or discarded. (The final statement in Groovy automatically becomes the return value, but you can also have explicit returns.)

File: script-common/src/test/java/org/opentestsystem/rdw/ingest/script/ScriptXmlTest.java (hunk 7 above)
Comment (gregcharles) [reply script-common/src/test/java/org/opentestsystem/rdw/ingest/script/ScriptXmlTest.java]: I'll see what I can do. I copied these tests, with some tweaks, from the ExamProcessorTest, and they include two assertThat methods from different classes. Obviously, only one can be statically imported. I'll try to figure out something clever to make that less ugly though.

File: script-common/build.gradle (hunk 8 above)
Comment (gregcharles) [reply script-common/build.gradle]: Yes, you're good. It looks like I just need it for a Hamcrest matcher (in org.hamcrest.text)
that it includes, but that my other dependencies don't. I'll either just bring in that library, or just implement my own matcher.

File: script-common/README.md (hunk 5 above)
Comment (malaffoon) [reply script-common/README.md]: Sorry, my comment wasn't clear. I'm thinking of conditional transform, perhaps something like:
```
transform '//Movie' { it.title = it.title.toUppercase() } 
when { it.title == 'Jaws' }
```

File: script-common/src/test/java/org/opentestsystem/rdw/ingest/script/ScriptXmlTest.java (hunk 7 above)
Comment (gregcharles) [reply script-common/src/test/java/org/opentestsystem/rdw/ingest/script/ScriptXmlTest.java]: I stand corrected. You can statically import methods with the same name from two different classes, I guess as long as the signatures are different. You learn something every day!

File: script-common/src/main/groovy/org/opentestsystem/rdw/ingest/script/DSLScriptBase.groovy (hunk 6 above)
Comment (gregcharles) [reply script-common/src/main/groovy/org/opentestsystem/rdw/ingest/script/DSLScriptBase.groovy]: Yes, I had a lot of trouble with this one. There are, in fact, builtin XML unescape/escape methods and they are called automatically when parsing and formatting the XML. What seems to be happening is that TI elements in the TRT have escaped XML in their text content, which is unescaped by the parser. However the XML has a <value> tag that itself contains an escaped HTML table definition in it. That means the html is doubly escaped, and so after parsing is still singly escaped. That would actually be fine, except the transform rule then tosses out the rest of the XML and just keeps the HTML part.  

tldr; It's sort of a cluster. I changed the name of this method to unescapeHtmlTags(), but I'll keep thinking about it.
